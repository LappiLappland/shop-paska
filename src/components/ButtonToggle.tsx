import { ReactNode, useRef } from 'react';
import useRipple from '../hooks/useRipple';

interface ButtonToggleProps {
  children?: ReactNode;
  className?: string;
  rippleClassName?: string;
  type?: 'submit' | 'reset' | 'button';
  active?: boolean;
  disabled?: boolean;
  dataTestid?: string;
  onClick?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

export default function ButtonToggle({
  children,
  className = '',
  rippleClassName = '',
  type,
  active = false,
  disabled,
  dataTestid,
  onClick,
}: ButtonToggleProps) {
  const buttonRef = useRef<HTMLButtonElement>(null);

  const ripple = useRipple(buttonRef, {
    color: 'bg-primary/focused',
    className: rippleClassName,
  });

  const colorsClassName = !active
    ? `
    border-outline text-on-surface
    hover:bg-secondary/hovered-r
    focus-visible:bg-secondary/focused-r
    active:bg-secondary/pressed-r
  `
    : `
    border-outline 
    text-on-secondary-container fill-on-secondary-container stroke-on-secondary-container
    bg-secondary-container
  `;

  return (
    <button
      ref={buttonRef}
      data-testid={dataTestid}
      className={`
      relative border
      transition-colors
      ${colorsClassName}
      ${className}`}
      type={type}
      disabled={disabled}
      onClick={(e) => {
        if (onClick) onClick(e);
      }}
    >
      {children}
      {ripple}
    </button>
  );
}
