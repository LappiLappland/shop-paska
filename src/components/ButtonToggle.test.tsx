import { act, render } from '@testing-library/react';
import ButtonToggle from './ButtonToggle';

describe('ButtonToggle component', () => {
  test('renders button toggle with default state', () => {
    const { container } = render(<ButtonToggle>Toggle</ButtonToggle>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders button toggle as active', () => {
    const { container } = render(
      <ButtonToggle active>Toggle</ButtonToggle>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders button toggle as disabled', () => {
    const { container } = render(
      <ButtonToggle disabled>Toggle</ButtonToggle>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('calls onClick handler when button toggle is clicked', () => {
    const onClick = jest.fn();
    const { getByRole } = render(
      <ButtonToggle onClick={onClick}>Toggle</ButtonToggle>
    );
    const button = getByRole('button');
    act(() => {
      button.click();
    })
    expect(onClick).toHaveBeenCalled();
  });
});