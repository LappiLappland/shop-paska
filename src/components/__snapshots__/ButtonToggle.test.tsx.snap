// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonToggle component renders button toggle as active 1`] = `
<button
  class="
      relative border
      transition-colors
      
    border-outline 
    text-on-secondary-container fill-on-secondary-container stroke-on-secondary-container
    bg-secondary-container
  
      "
>
  Toggle
  <span
    class="absolute inset-0 h-full w-full overflow-hidden "
  />
</button>
`;

exports[`ButtonToggle component renders button toggle as disabled 1`] = `
<button
  class="
      relative border
      transition-colors
      
    border-outline text-on-surface
    hover:bg-secondary/hovered-r
    focus-visible:bg-secondary/focused-r
    active:bg-secondary/pressed-r
  
      "
  disabled=""
>
  Toggle
  <span
    class="absolute inset-0 h-full w-full overflow-hidden "
  />
</button>
`;

exports[`ButtonToggle component renders button toggle with default state 1`] = `
<button
  class="
      relative border
      transition-colors
      
    border-outline text-on-surface
    hover:bg-secondary/hovered-r
    focus-visible:bg-secondary/focused-r
    active:bg-secondary/pressed-r
  
      "
>
  Toggle
  <span
    class="absolute inset-0 h-full w-full overflow-hidden "
  />
</button>
`;
